src/core/ucc_context.c

```diff
@@ -204,3 +205,84 @@ void ucc_context_config_print(const ucc_context_config_h config, FILE *stream,
             config->lib->full_prefix, (ucc_config_print_flags_t)flags);
     }
 }
+
+ucc_status_t ucc_context_create(ucc_lib_h lib,
+                                const ucc_context_params_t *params,
+                                const ucc_context_config_h  config,
+                                ucc_context_h *context)
+{
+    ucc_cl_context_params_t cl_ctx_params;
+    ucc_cl_context_t       *cl_ctx;
+    ucc_cl_lib_t           *cl_lib;
+    ucc_context_t          *ctx;
+    ucc_status_t            status;
+    uint64_t                i;
+    int                     num_cls;
+
+    num_cls = config->n_cl_cfg;
+    ctx     = ucc_malloc(sizeof(ucc_context_t), "ucc_context");
+    if (!ctx) {
+        ucc_error("failed to allocate %zd bytes for ucc_context",
+                  sizeof(ucc_context_t));
+        status = UCC_ERR_NO_MEMORY;
+        goto error;
+    }
+    ctx->lib = lib;
+    memcpy(&ctx->params, params, sizeof(ucc_context_params_t));
+
+    ctx->cl_ctx = (ucc_cl_context_t **)ucc_malloc(
+        sizeof(ucc_cl_context_t *) * num_cls, "cl_ctx_array");
+    if (!ctx->cl_ctx) {
+        ucc_error("failed to allocate %zd bytes for cl_ctx array",
+                  sizeof(ucc_cl_context_t *) * num_cls);
+        status = UCC_ERR_NO_MEMORY;
+        goto error_ctx;
+    }
+    ctx->n_cl_ctx        = num_cls;
+    cl_ctx_params.params = params;
+
+    for (i = 0; i < num_cls; i++) {
+        cl_lib = config->configs[i]->cl_lib;
+        status = cl_lib->iface->context_create(cl_lib, &cl_ctx_params,
+                                               config->configs[i], &cl_ctx);
+        if (UCC_OK != status) {
+            ucc_error("failed to create cl context for %s",
+                      cl_lib->iface->super.name);
+            goto error_ctx_create;
```

Comment:
Is it a fatal error and we cannot proceed? I would rather consider to print warn and continue and if no contexts were opened after the loop then print error

Reply:
I agree, doesn't seem fair to have one CL problem effect the others.

Reply:
@Sergei-Lebedev @alex--m I agree with you. One corner case: what if user specified the list of CLs explicitly? - e.g, basic,opt and, say, opt can not create ctx - is it an error then, or still just warn?

In case list of CLS is NOT specified, do we need to print warn or just "info" and proceed w/o additional output?

Reply:
I think if user explicitly sets CL list then print warning for each CL that cannot be created. If user doesn't set CL list all failed CL simply ignored. In both cases print error if no CL opened.

Reply:
I'd even go further than Sergei and say if user specified - failure should abort. If not - warn and continue...

Reply:
What is "abort" ? I guess returning not OK from ucc_context_create is fatal error from our perspective. But still user should decide how to process it: OMPI may fallback to another coll component ,for example.

Reply:
by abort I mean stop running (because we can't fill the user's request), which is what seems to happen today (but not what Sergei suggested).

Reply:
Seems like both Alex and Sergey have captured the issue very well.